### DISCUSSION TITLE
Поиск альбомов по ключевым словам

### DISCUSSION THREAD

**Discussion by amd64fox** [Chimildic/goofy#58] [Идея] [answered]
Последнее время очень плотно подсел на поиск плейлистов по ключевому слову, очень много всего можно найти там, зависит как поставить запрос. Параллельно искал в ручном режиме по альбомам, через десктопный клиент spotify и пришел к выводу, что и там и там много всякой интересной тематики, но есть различия, в альбомах нет такого хаоса как в плейлистах, все таки большинство плейлистов составлены спустя рукава, а альбомы в этом плане чуть более упорядочены и стабильны. Поэтому предлагаю рассмотреть вариант возвращения массива треков, найденных при поиске альбомов по ключевым словам.

**Comment by ilyanikolaich** [comment]
>> найденных при поиске альбомов по ключевым словам.

А как например, это может выглядеть? Слабо себе представляю поиск по альбомам - по ключевым словам. к Примеру поиск slippery when wet - безусловно сразу же приведет к альбому bon jovi, но вместе с этим и вывалит портянку альбомов, слабо кореллирующихся с такой музыкой.
Ну, то есть пока не совсем понятна логика - как можно использовать поиск по альбомам по ключевым словам?

**Reply by amd64fox** [reply]
Ключевое слово на то и ключевое, как ты его напишешь то тебе и выдаст, все зависит от ваших способностей мыслить нестандартно, названия альбома составляются с некоторыми  другими принципами нежели плейлисты,  при поиске это надо учитывать, например вбейте в поиск по альбомам `train sound`, и вы получите скоп альбомов связанных с звуками поезда, там так же будет схожи словосочетания с `train`, например `training` но в goofy такие исключения легко обработать и что самое важное это будет совершенно другой массив треков, нежели мы вобьём тоже самое в поиск по плейлистам, и там там свой охват определенных категорий и в этом я и считаю плюс данного метода поиска.

**Comment by amd64fox** [comment]
Или еще вариант для новой функции, поиск треков сразу по жанрам, например 
```javascript
let tracks = Source.getTracksgenre({
 genre: thrash metal
 year: 2021
 limit:   
});
```

**Reply by amd64fox** [reply]
сейчас даже так можно искать через клиент споти вбивая похожую комбинацию в поисковую строку:
`genre:thrash metal, year:2021`

**Comment by Chimildic** [comment, marked as answer]
Реализовано в [версии 1.4.5](https://chimildic.github.io/goofy/#/changelog?id=Версия-145)

**Reply by amd64fox** [reply]
а как сделать что бы выдавало больше 3-х треков ? 
```javascript
let tracks = Source.mineTracks({
    type: 'track',
    keyword: ['genre:edm + year:2021'],
});
```

**Reply by Chimildic** [reply]
`itemCount`

**Reply by amd64fox** [reply]
Кстати почему то по разному выдает кол-во треков:

Если через функцию тянуть то будет всего 7 треков.
```javascript
let tracks = Source.mineTracks({
    type: 'track',
    itemCount: 100,
    keyword: ['genre:thrash metal + year:2021'],
});
```
А если так `genre:thrash metal, year:2021` поискать через поиска клиента, то выдаст больше 40

думал в последнем варианте много дублей, но оказалось что нет

**Reply by Chimildic** [reply]
Мне при `genre:edm + year:2021` пришло 50 треков, 3 их которых удалились по дубликатам. Ручным поиском тоже много в результате.

Чтобы получить 100 элементов, увеличьте значение `requestCount`. По умолчанию 1, что даст только 50 элементов. Также не забывайте о `inRow`.

Возможно, Spotify не объединяет жанр и год в таком запросе `genre:thrash metal, year:2021`.

**Reply by amd64fox** [reply]
Ну вот сделал тоже самое для жанра `edm`
```javascript
let tracks = Source.mineTracks({
    type: 'track',
    itemCount: 1000,
    requestCount: 20,
    inRow: true,
    keyword: ['genre:edm + year:2021'],
});
```
Максимум, что удалось вытянуть через goofy это 564 трека.

Если искать через поиск клиента тоже самое `genre:edm, year:2021`, то выдает 1000 треков (похоже на ограничение), удаляю дубликаты, остается 877

При запросе из клиента, проверил что за треки он выдает, большее кол-во треков жанр edm, остальные 10-15% очень близкие жанры к edm, год у всех 2021.

**Reply by Chimildic** [reply]
В ответе на запрос поиска `genre:edm + year:2021` приходит значение `total = 628`. После удаления дубликатов 564. Больше через API мы не получим. В коде нет ошибки. 
Для 2020 года забивается вся тысяча треков, после фильтра 883.

**Reply by amd64fox** [reply]
Понял, говно-API 😁

**Reply by Chimildic** [reply]
Запрос `genre:edm, year:2021` не объединяет жанр и год. Если верить [документации расширенного поиска](https://support.spotify.com/by-ru/article/search/), объединение происходит через слово `AND` или знак `+`.
Поэтому запрос с запятой не ограничивается одним жанром или одним годом и забивается под 1 тысячу.

**Reply by Chimildic** [reply]
От 2021 года прошло мало. Возможно в этом жанре нет стольких треков, доступных нам. Хотя в запросе и нет статичной страны.

**Reply by ScreNN** [reply]
Хотел с помощью данной функции, собрать плейлист по жанру. В итоге после отработки кода
```
function example() {

let tracks = Source.mineTracks({
    type: 'track',
    keyword: ['genre:dubstep + year:2022'],
});

  console.log('Число запросов', CustomUrlFetchApp.getCountRequest());
  console.log('Треков в жанре Dubstep:', tracks.length);
}
```
**Возвращается значение Треков в жанре Dubstep: 3**
Как бы я не менял параметры, треков все равно возвращается 3.

Если взять в пример такой код
```
let tracks = Source.mineTracks({
    type: 'track',
    itemCount: 1000,
    requestCount: 20,
    inRow: true,
    keyword: ['genre:dubstep + year:2022'],
});
```
То на выходе получаем
**Треков в жанре Dubstep: 214**

Заходим в плейлист, слушаем, а там вообще каша какая то из жанров. Из жанра Dubstep там вообще по моему ничего нету.

**Reply by Chimildic** [reply]
@ScreNN Касаемо тройки, прочитайте [описание функции](https://chimildic.github.io/goofy/#/reference/source?id=minetracks). Это значение по умолчанию, когда `itemCount` явно не указан.

Проблема жанра в отвратительной разметки от Spotify. Во первых, известен только жанр в целом исполнителя, а не трека или альбома. Во вторых, даже у исполнителя вообще может не быть разметки. В третьих, качество и точность указанных жанров порой весьма субъективны. Некоторые исполнители вообще жанры меняют или сильно миксуют. Если руками напишите в поиске Spotify `genre:dubstep + year:2022`, то увидите то же самое. Возможно с некоторым различием в виду разницы между внутренним и внешним API.

Попробуйте использовать ключевое слово `dubstep` при типе `playlist`. В таком варианте мы смотрим на плейлисты явно собранные Spotify, кураторами или обычными пользователями, которые решили указать искомое слово в названии или описании. 

При этом _советую_ не зацикливаться на жанрах. Как в целом при использовании goofy, так и в частности этом примере. Куда точнее может быть поиск по ключевым словам, например, настроения или занятия. Такие явления очень посредственно отражаются через жанр. Особенно при плохой разметке. Но весьма точны когда плейлист собирается человеком или алгоритмом с конкретной целью.

**Reply by ScreNN** [reply]
Да я на жанре не зациклен, я слушаю любое направление, главное чтобы трек нравился. А из какого он жанра, неважно.
Просто иногда охота собрать плейлист определенного жанра, включить его и не заморачиваться.

**Reply by ScreNN** [reply]
В общем, попробовал вот так

```
function example11() {
    let tracks = Source.mineTracks({
        keyword: ['dubstep'],
        itemCount: 10,
        inRow: true,
        popularity: 50,
    });

    console.log('Число запросов', CustomUrlFetchApp.getCountRequest());
    console.log('Треков в жанре Dubstep:', tracks.length);

    let updTime = new Date().toLocaleString('ru-RU');

    Playlist.saveWithReplace({
        // id: '',
        name: 'Треки в жанре Dubstep',
        description: Playlist.getDescription(tracks)+'. '+ 'Дата обновления:' +' ' +updTime,
        tracks: tracks,
        randomCover: 'update',
        public: false,
    });
}
```
Получилось 93 трека, это я еще популярность фильтровал. Если без популярности, то там 1200 треков +/-.
Посмотрел на созданный плейлист, действительно, вижу что исполнители и треки из жанра Dubstep.

Теперь нужно понять что произошло.
`keyword: ['dubstep']` - найти плейлисты в названии которых присутствует ключевое слово "Dubstep"?
`itemCount: 10` - отобрать 10 плейлистов из всех найденных?
`popularity: 50` - Ну тут вроде ясно, популярность треков. Только не понятно как она считается. Найти все треки, у которых популярность от 50, во всех 10 найденных плейлистах? И отдать именно эти треки?

Еще бы примерно понимать, какое значение в итоге применяется к популярности. То есть 50 - это сколько примерно в прослушиваниях? Где то проскакивало аж 20 миллионов прослушиваний у трека. Если брать популярность 70, то там вообще 3 трека всего находит скрипт :)

Добавил еще в скрипт `followers: { min: 50, max: 1000 },` чтобы отобрать еще и плейлисты с подписчиками. В итоге конечное количество треков уменьшилось с 93, до 60.

В итоге сейчас так
```
function example11() {
    let tracks = Source.mineTracks({
        keyword: ['dubstep'],
        followers: { min: 50, max: 1000 },
        itemCount: 10,
        inRow: true,
        popularity: 30,
    });

    console.log('Число запросов', CustomUrlFetchApp.getCountRequest());
    console.log('Треков в жанре Dubstep:', tracks.length);

    let updTime = new Date().toLocaleString('ru-RU');

    Playlist.saveWithReplace({
        // id: '',
        name: 'Треки в жанре Dubstep',
        description: Playlist.getDescription(tracks)+'. '+ 'Дата обновления:' +' ' +updTime,
        tracks: tracks,
        randomCover: 'update',
        public: false,
    });
}
```
Поправочка, с `followers: { min: 50, max: 1000 }` плохо работает. Плейлист превратился опять в кашу из непонятный жанров. Убрал и все стало более менее нормально, но все равно какой то Рэп проскакивает, и так далее.

**Reply by Chimildic** [reply]
> популярность треков. Только не понятно как она считается

[Описание параметров объектов](https://chimildic.github.io/goofy/#/reference/desc?id=Описание-параметров-объектов). Популярность считает Spotify. Выхода к числу прослушиваний нет.

> В итоге конечное количество треков уменьшилось с 93, до 60

Уберите ключ `inRow` и добавьте `requestCount`. Тогда плейлисты выбираются случайно, что даст разнообразия с каждого обновления.

> Плейлист превратился опять в кашу из непонятный жанров

Вы просто автоматизировали механическую работу, собрав кучу треков из поиска в один список. Все "левое" ни ваша ошибка, ни ошибка goofy, а логика авторов плейлистов. Нежелательные жанры можно подчистить через `ban_genres` у [rangeTracks](https://chimildic.github.io/goofy/#/reference/filter?id=rangetracks).

Возможно вам больше понравится результат от ключевого слова `the sound of dubstep`.

**Reply by ScreNN** [reply]
Ну вроде вот так не плохо получилось

```
function GetTracksGenres() {
    let excludeTracks = [];
    let savedTracks = Cache.read('SavedTracks.json');
    let bannedTracks = Cache.read('BannedTracks.json');
    let recentTracks = RecentTracks.get();
    Combiner.push(excludeTracks, savedTracks, bannedTracks, recentTracks);

    let tracksGenres = Source.mineTracks({
        keyword: ['dubstep', 'drum-and-bass'],
        itemCount: 10,
        requestCount: 5,
        popularity: 60,
    });

    Filter.dedupTracks(tracksGenres);
    Filter.removeTracks(tracksGenres, excludeTracks);
    Filter.matchOriginalOnly(tracksGenres);

    Filter.rangeTracks(tracksGenres, {
        artist: {
            ban_genres: ['rap', 'r&b', 'metal', 'anime', 'soul', 'blues', 'punk', 'classic'],
        },
    });

    console.log('Число запросов', CustomUrlFetchApp.getCountRequest());
    console.log('Треков в жанре Dubstep:', tracksGenres.length);

    let updTime = new Date().toLocaleString('ru-RU');

    Playlist.saveWithReplace({
        // id: '',
        name: 'Треки в жанре Dubstep',
        description: Playlist.getDescription(tracksGenres)+'. '+ 'Дата обновления:' +' ' +updTime,
        tracks: tracksGenres,
        randomCover: 'update',
        public: false,
    });
}
```

**Reply by Chimildic** [reply]
Обратите внимание на флаг `isRemoveUnknownGenre` у _rangeTracks_. Часто у малопопулярных исполнителей нет разметки на жанры.

**Reply by ScreNN** [reply]
Вот как раз о нем я и думал. Хотел добавить. Спасибо
